sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -389,6 +389,46 @@ case class WriteDelta(
   }
 }
 
+trait V2CreateTableAsSelectPlan extends Command with V2CreateTablePlan with KeepAnalyzedQuery {
+  def name: LogicalPlan
+  def query: LogicalPlan
+  def isQueryAnalyzed: Boolean
+
+  override lazy val resolved: Boolean = childrenResolved && {
+    // the table schema is created from the query schema, so the only resolution needed is to check
+    // that the columns referenced by the table's partitioning exist in the query schema
+    val references = partitioning.flatMap(_.references).toSet
+    references.map(_.fieldNames).forall(query.schema.findNestedField(_).isDefined)
+  }
+
+  override def children: Seq[LogicalPlan] = if (isQueryAnalyzed) Seq(name) else Seq(name, query)
```

Comment:
The primary purpose of this trait is to share common logic for hiding `query` after the analysis to avoid optimizing the plan twice.